api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java

```diff
@@ -0,0 +1,139 @@
+package org.pmiops.workbench.api;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.inject.Provider;
+
+import org.pmiops.workbench.config.WorkbenchConfig;
+import org.pmiops.workbench.exceptions.ExceptionUtils;
+import org.pmiops.workbench.exceptions.ServerErrorException;
+import org.pmiops.workbench.model.CheckClustersResponse;
+import org.pmiops.workbench.notebooks.ApiException;
+import org.pmiops.workbench.notebooks.NotebooksConfig;
+import org.pmiops.workbench.notebooks.api.ClusterApi;
+import org.pmiops.workbench.notebooks.model.Cluster;
+import org.pmiops.workbench.notebooks.model.ClusterStatus;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.RestController;
+
+
+/**
+ * Offline cluster API. Handles cronjobs for cleanup and upgrade of older
+ * clusters. Methods here should be access restricted as, unlike
+ * ClusterController, these endpoints run as the Workbench service account.
+ */
+@RestController
+public class OfflineClusterController implements OfflineClusterApiDelegate {
+  private static final Logger log = Logger.getLogger(OfflineClusterController.class.getName());
+
+  private final Provider<ClusterApi> clusterApiProvider;
+  private final Provider<WorkbenchConfig> configProvider;
+  private final Clock clock;
+
+  @Autowired
+  OfflineClusterController(
+      @Qualifier(NotebooksConfig.SERVICE_CLUSTER_API) Provider<ClusterApi> clusterApiProvider,
+      Provider<WorkbenchConfig> configProvider,
+      Clock clock) {
+    this.clock = clock;
+    this.configProvider = configProvider;
+    this.clusterApiProvider = clusterApiProvider;
+  }
+
+  /**
+   * checkClusters deletes older clusters in order to force an upgrade on the
+   * next researcher login. This method is meant to be restricted to invocation
+   * by App Engine cron.
+   *
+   * The cluster deletion policy here aims to strike a balance between enforcing
+   * upgrades, cost savings, and minimizing user disruption. To this point, our
+   * goal is to only upgrade idle clusters if possible, as doing so gives us an
+   * assurance that a researcher is not actively using it. We delete clusters in
+   * the following cases:
+   *
+   * 1. It exceeds the max cluster age. Per environment, but O(weeks).
+   * 2. Since a cluster was last accessed, its spent over half its remaining
+   *    lifespan idle (as defined by [1]). This dynamic limit means that as the
+   *    upgrade horizon approaches, we'll be more aggressive in upgrading.
+   */
+  @Override
+  public ResponseEntity<CheckClustersResponse> checkClusters() {
```

Comment:
This logic is complicated enough (and the consequences dire enough if we get things wrong) that we probably should add some tests (mocking out ClusterApi).